beginner_source/scaled_dot_product_attention_tutorial.py

```diff
@@ -0,0 +1,322 @@
+"""
+Implement High-Performance Transformers with SCALED DOT PRODUCT ATTENTION
+================================================================================
+
+"""
+
+
+######################################################################
+# Summary
+# ~~~~~~~~
+#
+# In this tutorial, we want to highlight a new ``torch.nn.functional`` function
+# that can be helpful for implementing transformer architectures. The
+# function is named ``torch.nn.functional.scaled_dot_product_attention``.
+# For detailed description of the function, see the `PyTorch# documentation <https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html#torch.nn.functional.scaled_dot_product_attention>`__.
+# This function has already been incorporated into ``torch.nn.MultiheadAttention`` and ``torch.nn.TransformerEncoderLayer``.
+#
+# Overview
+# ~~~~~~~~~
+# At a high level, this PyTorch function calculates the
+# scaled dot product attention (SDPA) between query, key, and value according to
+# the definition found in the paper `Attention is all you
+# need <https://arxiv.org/abs/1706.03762>`__. While this function can be
+# written in PyTorch using existing functions, for GPU tensors this
+# function will implicitly dispatch to an optimized implementation. The
+# function is also highly modular and can be used to implement other
+# attention mechanisms such as
+# `Linformer <https://arxiv.org/abs/2006.04768>`__
+#
+# Fused implementations
+# ~~~~~~~~~~~~~~~~~~~~~~
+#
+# For CUDA tensor inputs, the function will dispatch into one of the following
+# implementations:
+#
+# * `FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness <https://arxiv.org/abs/2205.14135>`__
```

Comment:
Can you briefly explain the main innovations of each?

Reply:
I think the best summary of the innovations comes from the papers abstract:
"an IO-aware exact attention algorithm that uses tiling to reduce the number of memory reads/writes between GPU high bandwidth memory (HBM) and GPU on-chip SRAM." Honestly not sure the best way of abstracting this. 

memory-efficient attention is another implementation of the same fundamental algorithm that is more generic.